[24.0] Fix update group API payload model

davelopez opened the pull request:
All the payload fields should be optional when updating a group. Includes additional API test coverage for this.

Fixes https://github.com/usegalaxy-eu/gpu-access-sync/issues/1
```
Traceback (most recent call last):
  File "process.py", line 75, in <module>
    add_users()
  File "process.py", line 72, in add_users
    gc.update_group(group_id=gpu_group["id"], user_ids=l_user_ids, role_ids=[gpu_role[1]["id"]])
  File "/scratch/shiningpanda/jobs/e3616bad/virtualenvs/d41d8cd9/lib/python3.8/site-packages/bioblend/galaxy/groups/__init__.py", line 105, in update_group
    return self._put(payload=payload, id=group_id)
  File "/scratch/shiningpanda/jobs/e3616bad/virtualenvs/d41d8cd9/lib/python3.8/site-packages/bioblend/galaxy/client.py", line 165, in _put
    return self.gi.make_put_request(url, payload=payload, params=params)
  File "/scratch/shiningpanda/jobs/e3616bad/virtualenvs/d41d8cd9/lib/python3.8/site-packages/bioblend/galaxyclient.py", line 198, in make_put_request
    raise ConnectionError("Unexpected HTTP status code: %s" % r.status_code,
bioblend.ConnectionError: Unexpected HTTP status code: 400: {"err_msg":"Input should be a valid string in ('body', 'name')","err_code":400008,"validation_errors":["{\"type\": \"string_type\", \"loc\": [\"body\", \"name\"], \"msg\": \"Input should be a valid string\", \"input\": null}"]}
```

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Increase API test coverage for group update

The second update does not provide a group_name as we are not changing it. This will  fail with the current code as the update payload model is the same as the create one.

davelopez committed:
Fix update group payload model

All fields in the payload should be optional when updating.

davelopez committed:
Update client API schema

mvdbeek commented on lib/galaxy_test/api/test_groups.py:
```diff
@@ -76,10 +76,15 @@ def test_update(self):
+        another_user_id = None
+        with self._different_user():
+            another_user_id = self.dataset_populator.user_id()
+        assert another_user_id is not None
```
Do you need to reset the group members after the test ?

davelopez replied:
Hmmm... that could explain the interaction between tests... I keep forgetting about that, thanks! :sweat_smile: 

If that is the case, does this mean that adding a user (B) to a group where another user (A) is a member, will grant user B access to the datasets of user A? Or am I reading incorrectly the tests?

davelopez replied:
> does this mean that adding a user (B) to a group where another user (A) is a member, will grant user B access to the datasets of user A?

I'll answer myself, it is because the group also has the private role of user (A), so adding another user to the group will grant that user (B) access to the private role of user (A). Which probably makes sense and it is how permissions work.

mvdbeek commented on client/src/api/schema/schema.ts:
```diff
@@ -5483,6 +5483,15 @@ export interface components {
+            /** name of the group */
+            name?: string | null;
+            /** role IDs */
+            role_ids?: string[] | null;
```
I assume null should mean no modification should be made, that probably needs to be reflected in the code ?

davelopez replied:
So, I assume no provided value in the payload (None or null) is "leave it untouched" and explicit provided empty list means "remove all associations". Is that so?

I guess we need to change the old API behavior, it seems this `delete_existing_assocs=False` was preventing removing associations, only new ones could be added, right? :thinking: 

https://github.com/galaxyproject/galaxy/blob/d2956695024812b21898869f3f2127e882365f55/lib/galaxy/managers/groups.py#L95

mvdbeek replied:
>  explicit provided empty list means "remove all associations". Is that so?

That's a good question, I guess removing associations by omission is a bit of a breaking change 🤔. I'm sorry I missed your question yesterday.

> I guess we need to change the old API behavior, it seems this delete_existing_assocs=False was preventing removing associations

It would on the other hand make it so that not providing the role_ids parameter no change is made, which was my concern with https://github.com/galaxyproject/galaxy/pull/18374#discussion_r1636318636

davelopez replied:
No worries, I reverted the behavior back to keep all existing associations. The fact that once you add a user or role you cannot remove or replace it anymore still feels like a bug in the API and is not consistent with what the `manage_users_and_roles_for_group` admin controller endpoint does. But I understand this will be a breaking change.

I added more detailed checks to the tests to ensure nothing gets changed when it is not explicitly provided in the request.

davelopez committed:
Add tests for replacing and removing group users and roles

davelopez committed:
Fix update group API endpoint

Allow to replace or remove users and roles from the group

davelopez committed:
Fix typing

davelopez committed:
Keep old API behavior for now

Do not allow to remove or replace, only add users and roles.

davelopez committed:
Refactor API test for updating a group

davelopez committed:
Use random user in test

Since we cannot remove the user from the group to restore its previous state we need to use a user that will not be part of other tests to avoid unwanted interactions.
